File: backend/internal/models/passthrough.go
package models

import "time"

// PassthroughLog is one measured URL-passthrough relay. It records the usage
// the upstream itself reported (nil when it reports none); no cost.
type PassthroughLog struct {
	ID                 int64      `json:"id"`
	Host               string     `json:"host"`
	Path               string     `json:"path"`
	Method             string     `json:"method"`
	Model              string     `json:"model"`
	StatusCode         int        `json:"status_code"`
	IsError            bool       `json:"is_error"`
	ErrMessage         string     `json:"error_message,omitempty"`
	DNSMs              *int64     `json:"dns_ms"`
	ConnectMs          *int64     `json:"connect_ms"`
	TLSMs              *int64     `json:"tls_ms"`
	TTFBMs             *int64     `json:"ttfb_ms"`
	TTFTMs             *int64     `json:"ttft_ms"`
	TotalMs            int64      `json:"total_ms"`
	RequestBytes       int64      `json:"request_bytes"`
	ResponseBytes      int64      `json:"response_bytes"`
	InputTokens        *int64     `json:"input_tokens"`
	OutputTokens       *int64     `json:"output_tokens"`
	CacheWrite5MTokens *int64     `json:"cache_write_5m_tokens"`
	CacheWrite1HTokens *int64     `json:"cache_write_1h_tokens"`
	CacheReadTokens    *int64     `json:"cache_read_tokens"`
	StartedAt          time.Time  `json:"started_at"`
	CreatedAt          *time.Time `json:"created_at,omitempty"`
}

type PassthroughQueryParams struct {
	Host        string `form:"host"`
	From        string `form:"from"`
	To          string `form:"to"`
	Granularity string `form:"granularity"`
	Limit       int    `form:"limit"`
	Offset      int    `form:"offset"`
}

type PassthroughSummary struct {
	TotalRequests         int64    `json:"total_requests"`
	ErrorRate             float64  `json:"error_rate"`
	RequestsPerSec        float64  `json:"requests_per_sec"`
	AvgTotalMs            *float64 `json:"avg_total_ms"`
	P50TotalMs            *float64 `json:"p50_total_ms"`
	P95TotalMs            *float64 `json:"p95_total_ms"`
	P99TotalMs            *float64 `json:"p99_total_ms"`
	AvgTTFBMs             *float64 `json:"avg_ttfb_ms"`
	AvgTTFTMs             *float64 `json:"avg_ttft_ms"`
	AvgDNSMs              *float64 `json:"avg_dns_ms"`
	AvgConnectMs          *float64 `json:"avg_connect_ms"`
	AvgTLSMs              *float64 `json:"avg_tls_ms"`
	AvgResponseBytes      float64  `json:"avg_response_bytes"`
	TotalInputTokens      *int64   `json:"total_input_tokens"`
	TotalOutputTokens     *int64   `json:"total_output_tokens"`
	TotalCacheWrite5MToks *int64   `json:"total_cache_write_5m_tokens"`
	TotalCacheWrite1HToks *int64   `json:"total_cache_write_1h_tokens"`
	TotalCacheReadTokens  *int64   `json:"total_cache_read_tokens"`
}

type PassthroughBucket struct {
	Bucket           string   `json:"bucket"`
	RequestCount     int64    `json:"request_count"`
	ErrorCount       int64    `json:"error_count"`
	AvgTotalMs       *float64 `json:"avg_total_ms"`
	MaxTotalMs       int64    `json:"max_total_ms"`
	AvgTTFBMs        *float64 `json:"avg_ttfb_ms"`
	AvgResponseBytes float64  `json:"avg_response_bytes"`
	InputTokens      *int64   `json:"input_tokens"`
	OutputTokens     *int64   `json:"output_tokens"`
	CacheWrite5MToks *int64   `json:"cache_write_5m_tokens"`
	CacheWrite1HToks *int64   `json:"cache_write_1h_tokens"`
	CacheReadTokens  *int64   `json:"cache_read_tokens"`
}

type PassthroughHostStats struct {
	Host             string   `json:"host"`
	Requests         int64    `json:"requests"`
	Errors           int64    `json:"errors"`
	AvgTotalMs       *float64 `json:"avg_total_ms"`
	AvgTTFBMs        *float64 `json:"avg_ttfb_ms"`
	AvgTTFTMs        *float64 `json:"avg_ttft_ms"`
	AvgResponseBytes float64  `json:"avg_response_bytes"`
	OutputTokens     *int64   `json:"output_tokens"`
}

type PassthroughPerformanceResponse struct {
	Summary    PassthroughSummary     `json:"summary"`
	Timeseries []PassthroughBucket    `json:"timeseries"`
	ByHost     []PassthroughHostStats `json:"by_host"`
}

File: frontend/src/stores/passthrough.ts
import { defineStore } from "pinia";
import { ref } from "vue";
import api from "../api/client";

export interface PassthroughLog {
  id: number;
  host: string;
  path: string;
  method: string;
  model: string;
  status_code: number;
  is_error: boolean;
  error_message?: string;
  dns_ms: number | null;
  connect_ms: number | null;
  tls_ms: number | null;
  ttfb_ms: number | null;
  ttft_ms: number | null;
  total_ms: number;
  request_bytes: number;
  response_bytes: number;
  input_tokens: number | null;
  output_tokens: number | null;
  cache_write_5m_tokens: number | null;
  cache_write_1h_tokens: number | null;
  cache_read_tokens: number | null;
  started_at: string;
  created_at?: string;
}

export interface PassthroughSummary {
  total_requests: number;
  error_rate: number;
  requests_per_sec: number;
  avg_total_ms: number | null;
  p50_total_ms: number | null;
  p95_total_ms: number | null;
  p99_total_ms: number | null;
  avg_ttfb_ms: number | null;
  avg_ttft_ms: number | null;
  avg_dns_ms: number | null;
  avg_connect_ms: number | null;
  avg_tls_ms: number | null;
  avg_response_bytes: number;
  total_input_tokens: number | null;
  total_output_tokens: number | null;
  total_cache_write_5m_tokens: number | null;
  total_cache_write_1h_tokens: number | null;
  total_cache_read_tokens: number | null;
}

export interface PassthroughBucket {
  bucket: string;
  request_count: number;
  error_count: number;
  avg_total_ms: number | null;
  max_total_ms: number;
  avg_ttfb_ms: number | null;
  avg_response_bytes: number;
  input_tokens: number | null;
  output_tokens: number | null;
  cache_write_5m_tokens: number | null;
  cache_write_1h_tokens: number | null;
  cache_read_tokens: number | null;
}

export interface PassthroughHostStats {
  host: string;
  requests: number;
  errors: number;
  avg_total_ms: number | null;
  avg_ttfb_ms: number | null;
  avg_ttft_ms: number | null;
  avg_response_bytes: number;
  output_tokens: number | null;
}

export interface PassthroughPerformance {
  summary: PassthroughSummary;
  timeseries: PassthroughBucket[];
  by_host: PassthroughHostStats[];
}

export const usePassthroughStore = defineStore("passthrough", () => {
  const perf = ref<PassthroughPerformance | null>(null);
  const logs = ref<PassthroughLog[]>([]);
  const logTotal = ref(0);
  const loading = ref(false);
  const logsLoading = ref(false);
  const error = ref<string | null>(null);

  async function fetchPerformance(params?: Record<string, unknown>) {
    loading.value = true;
    error.value = null;
    try {
      const { data } = await api.get("/passthrough/performance", { params });
      perf.value = data as PassthroughPerformance;
    } catch (err: any) {
      error.value = err?.response?.data?.error || err?.message || "Failed to load relay metrics";
    } finally {
      loading.value = false;
    }
  }

  async function fetchLogs(params?: Record<string, unknown>) {
    logsLoading.value = true;
    try {
      const { data } = await api.get("/passthrough/logs", { params });
      logs.value = (data.data ?? []) as PassthroughLog[];
      logTotal.value = data.total ?? 0;
    } catch (err: any) {
      error.value = err?.response?.data?.error || err?.message || "Failed to load relay log";
    } finally {
      logsLoading.value = false;
    }
  }

  function clearError() {
    error.value = null;
  }

  return {
    perf,
    logs,
    logTotal,
    loading,
    logsLoading,
    error,
    fetchPerformance,
    fetchLogs,
    clearError,
  };
});
